Remove vtk8 support

biagas opened the pull request:
### Description

Remove VTK 8 support from build_visit, xml2cmake, source and CMake .

Resolves #19450


### Type of change

<!-- Please check one of the boxes below -->

* ~~[ ] Bug fix~~
* ~~[ ] New feature~~
* ~~[ ] Documentation update~~
* [X] Other
    Code cleanup

### How Has This Been Tested?

Ran build_visit on pascal, compiled against generated TP libs, ran full regression suite with success.

Compiled on Windows with success.

### Checklist:

- [X] I have commented my code where applicable.
- ~~[ ] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- ~~[ ] I have added debugging support to my changes.~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
Remove VTK 8 support from source code.

biagas committed:
Remove VTK 8 version of volume plot.

biagas committed:
Move VolumeVTK9 to Volume

biagas committed:
Rename VolumeVTK9 to Volume.

biagas committed:
Remove vtk8 support from build_visit.

Includes removing old ospray support libs tbb, ispc, embree.

biagas committed:
Remove support for ospray < 3.0 (tied to vtk8)

biagas committed:
Remove support for MSVC2017.

biagas committed:
Remove VTK 8 support OpenFOAM reader, molecule plot.

biagas committed:
Remove '--vtk9' usage.

biagas committed:
Re-add comment regarding cmake build doing the installation part as well.

biagas committed:
Fix typo

biagas committed:
Remove embree, ispc, tbb from modules.xml

biagas committed:
Ensure vtk lib version appended when running against install.

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

JustinPrivitera commented on src/CMake/SetUpThirdParty.cmake:
```diff
@@ -573,14 +571,12 @@ endif()
-    include(${VISIT_SOURCE_DIR}/CMake/FindPySide.cmake)
-  endif()
-endif()
+#if(VISIT_PYTHON_SCRIPTING AND PYTHONLIBS_FOUND
+#                          AND NOT VISIT_DBIO_ONLY
+#                          AND NOT VISIT_ENGINE_ONLY
+#                          AND NOT VISIT_SERVER_COMPONENTS_ONLY)
+#    include(${VISIT_SOURCE_DIR}/CMake/FindPySide.cmake)
+#endif()
```
Why commented out? Do we need this still?

biagas replied:
The if test for VTK 8 basically meant it was commented out for VTK 9. Probably should have an if-test based on Qt/Python versions instead. Can remove it and add a new ticket to see if we want to revive pyside.

biagas replied:
Created an issue, and added a new comment. Is this sufficient?

biagas committed:
Update SetUpThirdParty.cmake

Add reason for having FindPySide  commented out.

JustinPrivitera commented on src/doc/dev_manual/UpdatingPlugins.rst:
```diff
@@ -45,7 +45,7 @@ Tag                Purpose                                Supported components
 
 *xml2cmake* also supports the *Conditional* addition of an include directory, definition, link library or source file.
-The condition is a CMake variable that describes something related to the build: e.g. an OS-specification (*WIN32, LINUX, MACOS*), the availability of a third-party library (*HAVE_CONDUIT*) or a build option that can be toggled on/off (*VISIT_OSPRAY*). 
+The condition is a CMake variable that describes something related to the build: e.g. an OS-specification (*WIN32, LINUX, MACOS*), the availability of a third-party library (*HAVE_CONDUIT*) or a build option that can be toggled on/off. 
```
Do we want a new example here now that `(*VISIT_OSPRAY*)` was deleted?

biagas replied:
Good idea, I added `VISIT_PARALLEL` in its place.

biagas committed:
Update UpdatingPlugins.rst

Add a toggle option to replace VISIT_OSPRAY.

JustinPrivitera commented on src/viewer/main/viewer.C:
```diff
@@ -22,11 +22,7 @@
-#else
 #include <QVTKOpenGLNativeWidget.h>
-#endif
 
```
can we remove `#include <visit-config.h>` from the top of this file?

JustinPrivitera commented on src/visit_vtk/full/vtkUnstructuredGridRelevantPointsFilter.C:
```diff
@@ -7,9 +7,7 @@
 #include <visit-config.h>
```
can we remove `#include <visit-config.h>`

biagas committed:
Remove visit-config.h

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

JustinPrivitera commented on src/plots/Volume/avtVolumeFilter.C:
```diff
@@ -96,16 +87,17 @@ avtVolumeFilter::avtVolumeFilter()
 avtVolumeFilter::~avtVolumeFilter()
 {
-    if (primaryVariable != NULL)
+    if (primaryVariable != nullptr)
```
:)

biagas replied:
I just realized a note of explanation might be useful re: Volume plot.
`Volume` used to hold the code for the VTK-8 version of the plot.
`VolumeVTK` held the code for the VTK-9 version.
There was an if-test in plots/CMakeLists.txt to choose which one was built.

I deleted `Volume`, committed. Then renamed `VolumeVTK9` to `Volume`, but didn't make any changes.
In hindsight, this probably could have been handled in a separate PR to make things easier, my apologies.

JustinPrivitera replied:
No that's fine, I caught on after a bit.

I was reacting with the :) here because I was happy you changed from `NULL` to `nullptr`. A C++ instructor once told me you should be able to change every `NULL` to `nullptr` in your code, and if that causes a problem, then you've found a bug.

biagas replied:
So, do I need to address any of your new Volume plot comments?

JustinPrivitera replied:
Ohh that's where they're from. Sorry; sometimes I was just running through code and didn't realize it was from the volume changes.

You're, fine; I'll make them resolved.

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

JustinPrivitera commented on src/plots/Volume/QvisVolumePlotWindow.C:
```diff
@@ -278,9 +274,9 @@ QvisVolumePlotWindow::CreateWindowContents()
     // line edit widgets.
 #if (QT_VERSION >= QT_VERSION_CHECK(5, 11, 0))
-    int maxWidth = fontMetrics().horizontalAdvance("1.0000000000");
+    int maxWidth = fontMetrics().horizontalAdvance("0");
 #else
-    int maxWidth = fontMetrics().width("1.0000000000");
+    int maxWidth = fontMetrics().width("0");
 #endif
```
what is this change?

JustinPrivitera commented on src/plots/Volume/VolumeViewerEnginePluginInfo.C:
```diff
@@ -33,10 +33,12 @@ VolumeAttributes *VolumeViewerEnginePluginInfo::defaultAtts = NULL;
 //
 // ****************************************************************************
 
+#if 0
 #include <vtkRenderingVolumeOpenGL2ObjectFactory.h>
 
 static vtkRenderingVolumeOpenGL2ObjectFactory *volFactory = NULL;
 static int volFactory_count = 0;
+#endif
```
why was this change made?

JustinPrivitera commented on src/plots/Volume/VolumeViewerEnginePluginInfo.C:
```diff
@@ -46,13 +48,15 @@ VolumeViewerEnginePluginInfo::InitializeGlobalObjects()
         VolumeViewerEnginePluginInfo::clientAtts  = new VolumeAttributes;
         VolumeViewerEnginePluginInfo::defaultAtts = new VolumeAttributes;
     }
+#if 0
     // Want to ensure only 1 factory gets registered for this lib
     if (++volFactory_count == 1)
     {
         volFactory = vtkRenderingVolumeOpenGL2ObjectFactory::New();
         vtkObjectFactory::RegisterFactory(volFactory);
         volFactory->Delete();
     }
+#endif
```
same question here

JustinPrivitera commented on src/plots/Volume/VolumeViewerEnginePluginInfo.C:
```diff
@@ -216,9 +220,11 @@ VolumeViewerEnginePluginInfo::GetMenuName() const
 VolumeViewerEnginePluginInfo::~VolumeViewerEnginePluginInfo()
 {
     // Want to ensure only 1 factory gets unregistered for this lib
+#if 0
     if (--volFactory_count == 0)
     {
         vtkObjectFactory::UnRegisterFactory(volFactory);
     }
+#endif
```
and here

JustinPrivitera reviewed (dismissed):
This: ![image](https://github.com/visit-dav/visit/assets/35237779/9282ce1a-a637-40f5-8d2e-1eda946ac12e) makes me very happy. This looks like it was a ton of work, great job!

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas commented:
@markcmiller86 were you going to review this? If not I will merge as soon as the checks have finished.

markcmiller86 commented on src/CMake/FindVisItVTK.cmake:
```diff
@@ -21,12 +21,10 @@ if(NOT VTK_VERSION)
-    include(${VISIT_SOURCE_DIR}/CMake/FindVTK8.cmake)
 else()
-    message(FATAL_ERROR "Only VTK version 8.1.0, or >= 9.1.0 are supported")
+    message(FATAL_ERROR "Only VTK version >= 9.2.6 is supported")
```
Is the `>=` in this message likely to become misleading when VTK-10.0 is relased?

biagas replied:
Good point. Technically, we only support 9.2.6, though newer release of 9.2 might still work in theory, but I don't know if there will ever be a newer 9.2 release. Latest is 9.3 series and I believe there are API incompatibilities.
I will remove the `>=`.

biagas replied:
How about changing the whole block to this:
if(VTK_VERSION VERSION_GREATER_EQUAL "9.2.6" AND
   VTK_VERSION VERSION_LESS "9.3.0")
    include(${VISIT_SOURCE_DIR}/CMake/FindVTK9.cmake)
else()
    message(FATAL_ERROR "Only VTK version >= 9.2.6 and < 9.3.0 is currently supported")

markcmiller86 replied:
That works

markcmiller86 commented on src/CMake/PluginVsInstall.cmake.in:
```diff
@@ -196,7 +196,8 @@ set(VISIT_DBIO_ONLY              @VISIT_DBIO_ONLY@)
 set(VISIT_SLIVR                  @VISIT_SLIVR@)
-set(VISIT_OSPRAY                 @VISIT_OSPRAY@)
+set(HAVE_LIBOSPRAY               @HAVE_LIBOSPRAY@)
+set(HAVE_OSPRAY                  @HAVE_OSPRAY@)
```
Why do we need both of these ospray vars? In the code below, it looks like one is used for conditional logic in `CMakeLists.txt` files and one is used for conditional CPP logic in source code. However, I think we can name both the same, can't we? I asked ChatGTP...

### `CMakeLists.txt`
```
if (HAVE_LIBFOO)
    message(STATUS "Libfoo found")
    # Add libfoo related configuration here
else()
    message(STATUS "Libfoo not found")
    # Handle the absence of libfoo
endif()
```

### Passing to CPP source code
```
add_executable(MyApp main.c)

if (HAVE_LIBFOO)
    target_compile_definitions(MyApp PRIVATE HAVE_LIBFOO=1)
else()
    target_compile_definitions(MyApp PRIVATE HAVE_LIBFOO=0)
endif()
```

biagas replied:
You are correct, the logic for this could be handled differently, and probably with one CMake var.
Current CMake logic doesn't use target_compile_definitions for HAVE_OSPRAY like it should, but instead uses the global 'add_definitions' in the FindOSPRay module.  The need to support building PluginsVsInstall in its current form is the reason for adding both vars here.

I can enter a new ticket to fix this, or make it a bullet point in my cmake modernization efforts.
Which would you prefer?

markcmiller86 replied:
Ok, sounds fine.

markcmiller86 commented on src/plots/Volume/QvisVolumePlotWindow.C:
Are there substantial changes to Volume plot GUI due to changes in VTK?

biagas replied:
VTK -9 version of Volume plot was called VolumeVTK9, due to the massive changes that made it difficult to support both VTK 8 and 9 in the same source code.
I deleted Volume (the vtk 8 version), commited that change, then renamed VolumeVTK9 to Volume without making any changes to the code.
I apologize for not thinking about the repercussion for reviewers. It could have been done in a separate PR once the VTK-8 removal was finalized.

markcmiller86 replied:
No worries. Just wasn't sure why there were so many changes to the Volume plot for removing VTK 8

markcmiller86 commented on src/plots/Volume/avtLowerResolutionVolumeFilter.C:
```diff
@@ -7,17 +7,21 @@
+#include <vtkPointData.h>
+#include <vtkSkew.h>
+
+#define NO_DATA_VALUE -1e+37
```
This constant seems oddly named for how it is used. It is used in `<` and `>` comparisons. Is this a magic value used in other source files? Does it make sense to use `std::numeric_limits<float>::max()`

biagas replied:
This is code that Allen committed as part of the VTK 9 port, and was previously reviewed and approved.

markcmiller86 replied:
In your working on this code, did you find the use of this symbol confusing? What does it actually mean? I am not sure I can tell from either the name or the usage.

biagas replied:
I have no comments to make regarding the Volume plot. I did not work with the code, I only renamed from VolumeVTK9 after I removed Volume (vtk 8 version). If there are issues of concern with the Volume plot in its current form, may I suggest you create new issue(s).

markcmiller86 commented on src/plots/Volume/avtLowerResolutionVolumeFilter.h:
```diff
@@ -41,16 +41,20 @@ class avtLowerResolutionVolumeFilter : public avtPluginDataTreeIterator
     VolumeAttributes         atts;
-    float                   *hist;
-    int                      hist_size;
+    float                   *hist{nullptr};
```
I am unfamiliar with this `{value}` syntax here in a header file. Is this defining default values?

biagas replied:
Yes.

markcmiller86 commented on src/plots/Volume/avtVolumeRenderer.C:
```diff
@@ -7,26 +7,48 @@
+#include <vtkGeometryFilter.h>
+#include <vtkPolyDataMapper.h>
+
+#ifndef NO_DATA_VALUE
```
This copies a definition elsewhere and so am assuming both definitions need to be kept consistent. Just wondering if there is a better *single* place to define it for both so there is no chance of them getting out of sync. And, again IMHO the symbol name doesn't really match its use.

markcmiller86 commented on src/plots/Volume/avtVolumeRenderer.C:
```diff
@@ -135,374 +96,121 @@ avtVolumeRenderer::New(void)
-            gm_max = VolumeCalculateGradient(atts, grid, opac, gx, gy, gz, gm, gmn, ghostval);
+        for( int ptId = 0; ptId<nTupples; ++ptId )
+        {
+            double dataTuple = dataArr->GetTuple1(ptId);
```
`NO_DATA_VALUE`'s value is close to the largest negative 32 bit float. But, it is way far from the largest negative 64 bit double. So, this means that a large fraction of the total range of values representable by a double is considered out of bounds. Is this correct?

markcmiller86 commented on src/tools/dev/xml/GenerateCMake.h:
```diff
@@ -308,7 +303,10 @@ class CMakeGeneratorPlugin : public Plugin
+                    tmp.append(vtkversion);
                     libs9.push_back(tmp);
+                }
             }
```
I assume `libs9` and `libs8` variables were introduced to handle either VTK 8 or VTK 9. Any reason to get rid of the `9` designation in `libs9` now?

markcmiller86 commented on src/tools/dev/xml/GenerateCMake.h:
```diff
@@ -596,24 +591,9 @@ class CMakeGeneratorPlugin : public Plugin
             CMakeWrite_TargetLinkDirs(out, "", "E", "_ser", eldflagsSer);
         }
-        if (!vtk8_elibsSer.empty())
+        if (!vtk9_elibsSer.empty())
```
Is the `9` designation still needed on this variable?

biagas replied:
Probably not. This will all be rewritten with my CMake modernization efforts. Can it wait until then?

markcmiller86 replied:
Thats fine.

markcmiller86 reviewed (commented):
@biagas based on what I saw here, I think there is a small but nonetheless interesting software engineereing story to tell about how and why it was necessary to enable VisIt be built with both VTK 8 and VTK 9 for a transitional period. I think the work you needed to do to enable that, support it for the transition period and then remove it is a really good example of the kinds of critical software engineering activities are necessary in keeping up to date with dependencies. I think this would make a nice short article for bssw.io and I'd be happy to get you started with it.

Also, for the PR review, I made a number of comments and questions.

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

biagas committed:
Change VTK version checks.

We don't support 9.3 or higher currently.
Don't know if there will be a newer 9.2 release, but make allowances just in case.

biagas committed:
Merge branch 'develop' into task/biagas/remove_vtk8_support

markcmiller86 commented:
@biagas there were two unresponded/unreslolved comments here. Both relate, loosely though, to previous comments. I wasn't sure if you consider those resolved too or are still considering?

biagas commented:
> @biagas there were two unresponded/unreslolved comments here. Both relate, loosely though, to previous comments. I wasn't sure if you consider those resolved too or are still considering?

Since they were related to previous comments, I thought we were done with them.
